// src/Pages/BankDetails.js
import React, { useState, useEffect } from 'react';
import { Container, Row, Col, Form, Button, Card } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import axios from 'axios';

function BankDetails() {
    const [formData, setFormData] = useState({
        staffId: '',
        employeeName: '',
        bankName: '',
        accountNumber: '',
        accountName: '',
        phoneNumber: ''
    });

    const [bankDetailsList, setBankDetailsList] = useState([]);

    // Handle input changes
    const handleInputChange = (e) => {
        const { name, value } = e.target;
        setFormData({ ...formData, [name]: value });
    };

    // Fetch existing bank details from the backend (if applicable)
    useEffect(() => {
        fetchBankDetails();
    }, []);

    const fetchBankDetails = async () => {
        try {
            const response = await axios.get('http://localhost:5001/api/bankdetails');
            setBankDetailsList(response.data);
        } catch (error) {
            console.error("Error fetching bank details:", error);
        }
    };

    // Handle form submission to save bank details
    const handleSubmit = async (e) => {
        e.preventDefault();

        const bankData = {
            staffId: formData.staffId,
            employeeName: formData.employeeName,
            bankName: formData.bankName,
            accountNumber: formData.accountNumber,
            accountName: formData.accountName,
            phoneNumber: formData.phoneNumber
        };

        try {
            await axios.post('http://localhost:5001/api/bankdetails', bankData);
            alert('Bank details saved successfully');
            setFormData({
                staffId: '',
                employeeName: '',
                bankName: '',
                accountNumber: '',
                accountName: '',
                phoneNumber: ''
            });
            // Refetch the updated data
            fetchBankDetails();
        } catch (error) {
            console.error("Error saving bank details:", error);
        }
    };

    return (
        <Card className="m-3">
            <Card.Header>
                <h3 className="text-left">Bank Details Management</h3>
            </Card.Header>
            <Card.Body>
                <Container fluid>
                    <Form onSubmit={handleSubmit}>
                        <Row className="mb-3">
                            <Col md={4}>
                            <Form.Group controlId="employeeName">
                                    <Form.Label>Employee Name</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="employeeName"
                                        value={formData.employeeName}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                            <Col md={4}>
                            <Form.Group controlId="staffId">
                                    <Form.Label>Staff ID</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="staffId"
                                        value={formData.staffId}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                            <Col md={4}>
                                <Form.Group controlId="bankName">
                                    <Form.Label>Bank Name</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="bankName"
                                        value={formData.bankName}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                        </Row>

                        <Row className="mb-3">
                            <Col md={4}>
                                <Form.Group controlId="accountNumber">
                                    <Form.Label>Account Number</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="accountNumber"
                                        value={formData.accountNumber}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                            <Col md={4}>
                                <Form.Group controlId="accountName">
                                    <Form.Label>Account Name</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="accountName"
                                        value={formData.accountName}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                            <Col md={4}>
                                <Form.Group controlId="phoneNumber">
                                    <Form.Label>Phone Number</Form.Label>
                                    <Form.Control
                                        type="text"
                                        name="phoneNumber"
                                        value={formData.phoneNumber}
                                        onChange={handleInputChange}
                                        required
                                    />
                                </Form.Group>
                            </Col>
                        </Row>

                        <Button variant="primary" type="submit">
                            Save
                        </Button>
                    </Form>

                    <hr />

                    {bankDetailsList.length > 0 && (
                        <div className="mt-4">
                            <h4>Saved Bank Details</h4>
                            <div className="table-responsive">
                                <table className="table table-bordered table-striped">
                                    <thead>
                                        <tr>
                                            <th>Staff ID</th>
                                            <th>Employee Name</th>
                                            <th>Bank Name</th>
                                            <th>Account Number</th>
                                            <th>Account Name</th>
                                            <th>Phone Number</th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        {bankDetailsList.map((data, index) => (
                                            <tr key={index}>
                                                <td>{data.staffId}</td>
                                                <td>{data.employeeName}</td>
                                                <td>{data.bankName}</td>
                                                <td>{data.accountNumber}</td>
                                                <td>{data.accountName}</td>
                                                <td>{data.phoneNumber}</td>
                                            </tr>
                                        ))}
                                    </tbody>
                                </table>
                            </div>
                        </div>
                    )}
                </Container>
            </Card.Body>
        </Card>
    );
}

export default BankDetails;
