// models/Payroll.js
// const mongoose = require('mongoose');

// // Create the payroll schema
// const payrollSchema = new mongoose.Schema({
//   staffID: { type: String, required: true },
//   employeeName: { type: String, required: true },
//   bankName: { type: String, required: true },
//   accountNumber: { type: String, required: true },
//   accountName: { type: String, required: true },
//   monthlyEarnings: { type: Number, required: true },
//   totalDeductions: { type: Number, required: true },
//   netSalary: { type: Number, required: true },
// });

// // Create the Payroll model
// const Payroll = mongoose.model('Payroll', payrollSchema);

// module.exports = Payroll;

//------------------------------------------------------------------------------

// models/Payroll.js
const mongoose = require('mongoose');

const PayrollSchema = new mongoose.Schema({
  year: { type: Number, required: true},
  month: { type: String, required: true},
  staffID: { type: String, required: true },
  employeeName: { type: String, required: true },
  bankName: { type: String, required: true },
  accountNumber: { type: String, required: true },
  accountName: { type: String, required: true },
  monthlyEarnings: { type: Number, required: true },
  totalDeductions: { type: Number, required: true },
  netSalary: { type: Number, required: true },
}, { timestamps: true });

module.exports = mongoose.model('Payroll', PayrollSchema);
