// src/Componets/Employee.js
// Components/AllEmployees
import { useState } from "react";

export const useEmployeeDetails = () => {
  const [employees, setEmployees] = useState([
    {
      profile: "Jeremy Neigh",
      id: "A0B1C001",
      status: "Part-Time",
      department: "Support",
      shift: "8h-17h",
      joiningDate: "2024-10-18",
      role: "Tax Officer",
      basicSalary: "20000",
      approveLoan: "Yes",
      markingTime: "08:00",
    },
    {
      profile: "Emily Johnson",
      id: "A0B1C002",
      status: "Full-Time",
      department: "Finance",
      shift: "9h-18h",
      joiningDate: "2024-10-18",
      role: "Accountant",
      basicSalary: "4000",
      approveLoan: "No",
      markingTime: "09:00",
    },
    {
      profile: "Mark Smith",
      id: "A0B1C003",
      status: "On-contract",
      department: "IT",
      shift: "8h-17h",
      joiningDate: "2024-10-18",
      role: "Software Developer",
      basicSalary: "5000",
      approveLoan: "Yes",
      markingTime: "08:30",
    },
    {
      profile: "Sophia Williams",
      id: "A0B1C004",
      status: "Full-Time",
      department: "HR",
      shift: "8h-17h",
      joiningDate: "2024-10-18",
      role: "HR Manager",
      basicSalary: "4500",
      approveLoan: "No",
      markingTime: "08:00",
    },
    {
      profile: "Liam Brown",
      id: "A0B1C005",
      status: "Seasonal",
      department: "Marketing",
      shift: "10h-19h",
      joiningDate: "2024-10-18",
      role: "Marketing Specialist",
      basicSalary: "3000",
      approveLoan: "No",
      markingTime: "10:00",
    },
    {
      profile: "Olivia Garcia",
      id: "A0B1C006",
      status: "Full-Time",
      department: "Sales",
      shift: "8h-17h",
      joiningDate: "2024-10-18",
      role: "Sales Executive",
      basicSalary: "3500",
      approveLoan: "Yes",
      markingTime: "08:00",
    },
    {
      profile: "Noah Martinez",
      id: "A0B1C007",
      status: "Part-Time",
      department: "Operations",
      shift: "6h-15h",
      joiningDate: "2024-10-18",
      role: "Operations Coordinator",
      basicSalary: "2500",
      approveLoan: "Yes",
      markingTime: "06:00",
    },
  ]);

  return { employees, setEmployees };
};
